src/simsopt/solve/permanent_magnet_optimization.py

```diff
@@ -0,0 +1,415 @@
+        # convex algorithm called MwPGP
+        algorithm_history, _, m_history, m = convex_step(
+            A_obj=pm_opt.A_obj,
+            b_obj=pm_opt.b_obj,
+            ATb=ATb,
+            m_proxy=m0,
+            m0=m0,
+            m_maxima=mmax,
+            **kwargs_convex
+        )
+        m = np.ravel(m)
+        m_proxy = m
+
+    # note m = m_proxy if not using relax-and-split (i.e. problem is convex)
+    pm_opt.m = m
+    pm_opt.m_proxy = m_proxy
+    return errors, m_history, m_proxy_history
+
+
+def GPMO(pm_opt, algorithm='baseline', **algorithm_kwargs):
+    """
+    GPMO is a greedy algorithm alternative to
+    the relax-and-split algorithm for solving the permanent
+    magnet optimization problem. Full-strength magnets are placed
+    one-by-one according to minimize 
+    the MSE (fB). Allows for a number of keyword arguments that 
+    facilitate some basic backtracking (error correction) and 
+    placing magnets together so no isolated magnets occur.
+
+    Args:
+        pm_opt: PermanentMagnetGrid class object
+            The grid of permanent magnets to optimize.
+        algorithm: string
+            The type of greedy algorithm to use. Options are
+            baseline (the simple implementation of GPMO), multi
+            (place multiple magnets per iteration), backtracking
+            (backtrack every few hundred iterations to improve the
+            solution), ArbVec (the simple implementation of GPMO,
+            but with arbitrary oriented polarization vectors), and
+            ArbVec_backtracking (same as above but w/ backtracking).
+            Easiest to use is baseline but most effective algorithm
+            is ArbVec_backtracking.
+        algorithm_kwargs: dict
+            Keyword argument dictionary for the GPMO algorithm
+            and its variants.
+    """
+    if not hasattr(pm_opt, "A_obj"):
+        raise ValueError(
+            "The PermanentMagnetClass needs to use geo_setup() or "
+            "geo_setup_from_famus() before calling optimization routines."
+        )
+
+    # Begin the various algorithms
+    errors = []
+    m_history = []
+
+    # Need to normalize the m, so have
+    # || A * m - b||^2 = || ( A * mmax) * m / mmax - b||^2
+    mmax = pm_opt.m_maxima
+    contig = np.ascontiguousarray
+    mmax_vec = contig(np.array([mmax, mmax, mmax]).T.reshape(pm_opt.ndipoles * 3))
+    A_obj = pm_opt.A_obj * mmax_vec
+
+    if (algorithm != 'baseline' and algorithm != 'mutual_coherence' and algorithm != 'ArbVec') and 'dipole_grid_xyz' not in algorithm_kwargs.keys():
+        raise ValueError('GPMO variants require dipole_grid_xyz to be defined.')
+
+    # Run one of the greedy algorithm (GPMO) variants
+    if "reg_l2" in algorithm_kwargs.keys():
+        reg_l2 = algorithm_kwargs["reg_l2"]
+        algorithm_kwargs.pop("reg_l2")
+    else:
+        reg_l2 = 0.0
+
+    # check that algorithm can generate K binary dipoles
+    if "K" in algorithm_kwargs.keys():
```

Comment:
This is old school.
 
`if "K" in algorithm_kwargs:` should be sufficient. Noticed similar usage elsewhere.